File: src/ExamSolution.py
import math
import random
from pathlib import Path

from .Seat import Seat


class ExamSolution:
    """Represents one seating assignment. seating[i] is student i's seat."""

    def __init__(self, instance, random_solution: bool = False):
        self.instance = instance

        # seating[student_id] stores the room, row, and column assigned to that student.
        self.seating = [Seat(-1, -1, -1) for _ in range(instance.N)]

        if random_solution:
            seats = []

            # Collect all available seats.
            for room in instance.rooms:
                for row in range(room.getRows()):
                    for col in range(room.getCols()):
                        seats.append(Seat(room.getId(), row, col))

            if len(seats) < instance.N:
                raise ValueError("Not enough seats for all students.")

            random.shuffle(seats)

            # Assign seats to students 0..N-1.
            for student_id in range(instance.N):
                self.seating[student_id] = seats[student_id]

    def copy(self):
        """Returns an independent copy of this solution."""
        sol = ExamSolution(self.instance)
        sol.seating = [seat.copy() for seat in self.seating]
        return sol

    def seatStudent(self, student, seat: Seat):
        """Assigns one student to one seat."""
        self.seating[student.id] = seat

    def isValidSeatPosition(self, seat: Seat) -> bool:
        """Checks whether a seat is inside its room bounds."""
        if seat.getRoomId() < 0 or seat.getRoomId() >= self.instance.M:
            return False

        room = self.instance.rooms[seat.getRoomId()]

        return (
            0 <= seat.getRow() < room.getRows()
            and 0 <= seat.getCol() < room.getCols()
        )

    def checkValid(self) -> bool:
        """Checks that all students have valid, unique seats."""
        taken = set()

        for student_id in range(self.instance.N):
            seat = self.seating[student_id]

            if not self.isValidSeatPosition(seat):
                return False

            if seat in taken:
                return False

            taken.add(seat)

        return True

    def cheaterIsolation(self) -> int:
        """Counts same-group students adjacent to listed cheaters."""
        cheat_penalty = 0

        for cheater in self.instance.cheaters:
            cheater_seat = self.seating[cheater.id]
            group = self.instance.groups[cheater.group_id]

            for other in group.students:
                if cheater.id == other.id:
                    continue

                other_seat = self.seating[other.id]

                if (
                    cheater_seat.room_id == other_seat.room_id
                    and abs(cheater_seat.row - other_seat.row) <= 1
                    and abs(cheater_seat.col - other_seat.col) <= 1
                ):
                    cheat_penalty += 1

        return cheat_penalty

    def _seatsByRoom(self, group):
        """Groups the seats of one exam group by room."""
        distribution = {}

        for student in group.students:
            seat = self.seating[student.id]

            if seat.room_id not in distribution:
                distribution[seat.room_id] = []

            distribution[seat.room_id].append(seat)

        return distribution

    def _rawGroupSplittingCost(self) -> float:
        """Computes the group splitting part before normalization."""
        splitting_cost = 0.0

        for group in self.instance.groups:
            distribution = self._seatsByRoom(group)
            splitting_cost += len(distribution) * self.instance.cSplit

        return splitting_cost

    def _rawDiameterCost(self) -> float:
        """Computes the diameter part before normalization."""
        diameter_cost = 0.0

        for group in self.instance.groups:
            distribution = self._seatsByRoom(group)

            for seats in distribution.values():
                diam_squared = 0.0

                # Find this group's diameter in the room.
                for i in range(len(seats)):
                    seat1 = seats[i]

                    for j in range(i + 1, len(seats)):
                        seat2 = seats[j]

                        dist_squared = (
                            (seat1.col - seat2.col) ** 2
                            + (seat1.row - seat2.row) ** 2
                        )

                        diam_squared = max(diam_squared, dist_squared)

                diameter_cost += math.sqrt(diam_squared) * self.instance.cDiam

        return diameter_cost

    def invigilatorCost(self) -> float:
        """Computes group splitting and diameter costs."""
        return self._rawGroupSplittingCost() + self._rawDiameterCost()

    def _cheatPenaltyByRoom(self):
        """Computes the number of cheating penalties per room."""
        penalties = [0 for _ in range(self.instance.M)]

        for cheater in self.instance.cheaters:
            cheater_seat = self.seating[cheater.id]
            group = self.instance.groups[cheater.group_id]

            for other in group.students:
                if cheater.id == other.id:
                    continue

                other_seat = self.seating[other.id]

                if (
                    cheater_seat.room_id == other_seat.room_id
                    and abs(cheater_seat.row - other_seat.row) <= 1
                    and abs(cheater_seat.col - other_seat.col) <= 1
                ):
                    penalties[cheater_seat.room_id] += 1

        return penalties

    def _formatDouble(self, value: float) -> str:
        """Formats doubles without unnecessary trailing zeroes."""
        if abs(value) < 1e-9:
            value = 0.0

        result = f"{value:.6f}".rstrip("0").rstrip(".")
        return result if result and result != "-0" else "0"

    def getCost(self) -> float:
        """Computes the normalized objective value."""
        cheat_term = 0.0

        if len(self.instance.cheaters) > 0:
            cheat_term = (
                self.instance.cCheat / len(self.instance.cheaters)
            ) * self.cheaterIsolation()

        invig_term = 0.0

        if len(self.instance.groups) > 0:
            invig_term = (1.0 / len(self.instance.groups)) * self.invigilatorCost()

        return cheat_term + invig_term

    def stats(self, filename: str):
        """Writes a detailed cost breakdown to statistics/<filename>."""
        file_path = Path("statistics") / filename

        total_cheat_penalty = self.cheaterIsolation()
        room_cheat_penalty = self._cheatPenaltyByRoom()

        cheating_term = 0.0
        if len(self.instance.cheaters) > 0:
            cheating_term = (
                self.instance.cCheat / len(self.instance.cheaters)
            ) * total_cheat_penalty

        splitting_cost = self._rawGroupSplittingCost()
        diameter_cost = self._rawDiameterCost()
        invigilator_cost = splitting_cost + diameter_cost

        invigilator_term = 0.0
        if len(self.instance.groups) > 0:
            invigilator_term = invigilator_cost / len(self.instance.groups)

        total_cost = cheating_term + invigilator_term

        with open(file_path, "w", encoding="utf-8") as writer:
            writer.write("Cheater Isolation Statistics\n")
            writer.write("----------------------------\n")
            writer.write(f"Total cheatPenalty = {total_cheat_penalty}\n")
            writer.write(f"Number of cheaters |Q| = {len(self.instance.cheaters)}\n")
            writer.write(f"c_cheat = {self._formatDouble(self.instance.cCheat)}\n")
            writer.write(f"Normalized cheating term = {self._formatDouble(cheating_term)}\n\n")

            writer.write("Cheat penalty by room:\n")
            for room_id in range(self.instance.M):
                writer.write(
                    f"Room {room_id} cheatPenalty = {room_cheat_penalty[room_id]}\n"
                )
            writer.write("\n")

            writer.write("Invigilator Statistics\n")
            writer.write("----------------------\n")
            writer.write(f"Raw group splitting cost = {self._formatDouble(splitting_cost)}\n")
            writer.write(f"Raw diameter cost = {self._formatDouble(diameter_cost)}\n")
            writer.write(f"Raw invigilator cost = {self._formatDouble(invigilator_cost)}\n")
            writer.write(f"Number of groups G = {len(self.instance.groups)}\n")
            writer.write(
                f"Normalized invigilator term = {self._formatDouble(invigilator_term)}\n\n"
            )

            writer.write("Combined Objective\n")
            writer.write("------------------\n")
            writer.write(f"Cheating term = {self._formatDouble(cheating_term)}\n")
            writer.write(f"Invigilator term = {self._formatDouble(invigilator_term)}\n")
            writer.write(f"Total cost F = {self._formatDouble(total_cost)}\n")

    def output(self, filename: str):
        """Writes one line per student: room row col."""
        with open(filename, "w", encoding="utf-8") as file:
            for seat in self.seating:
                file.write(f"{seat.room_id} {seat.row} {seat.col}\n")

    def _colorFromGroup(self, group_id: int) -> str:
        """Returns a color for a group."""
        palette = [
            "#FFADAD", "#FFD6A5", "#FDFFB6", "#CAFFBF", "#9BF6FF",
            "#A0C4FF", "#BDB2FF", "#FFC6FF", "#B9FBC0", "#F1C0E8",
            "#CFBAF0", "#A3C4F3", "#90DBF4", "#98F5E1", "#FDE4CF",
            "#E4C1F9", "#D0F4DE", "#FCF6BD", "#FFCFD2", "#CDE7BE",
            "#FBC4AB", "#BDE0FE", "#CDB4DB", "#FFFFB5", "#D8F3DC",
        ]

        if 0 <= group_id < len(palette):
            return palette[group_id]

        hue = (group_id * 0.618033988749895) % 1.0
        saturation = 0.42
        brightness = 1.0

        return self._hsvToHex(hue, saturation, brightness)

    def _hsvToHex(self, h: float, s: float, v: float) -> str:
        i = int(h * 6)
        f = h * 6 - i
        p = v * (1 - s)
        q = v * (1 - f * s)
        t = v * (1 - (1 - f) * s)
        i %= 6

        if i == 0:
            r, g, b = v, t, p
        elif i == 1:
            r, g, b = q, v, p
        elif i == 2:
            r, g, b = p, v, t
        elif i == 3:
            r, g, b = p, q, v
        elif i == 4:
            r, g, b = t, p, v
        else:
            r, g, b = v, p, q

        return "#{:02X}{:02X}{:02X}".format(
            round(r * 255), round(g * 255), round(b * 255)
        )

    def visualize(self, filename: str):
        """Writes an SVG visualization to visualizations/<filename>."""
        filename_path = Path(filename)

        if filename_path.parent == Path("."):
            file_path = Path("visualizations") / filename_path.name
        else:
            file_path = filename_path

        cell = 64
        pad = 32
        title_h = 44
        room_gap = 42
        room_inner_pad = 18
        bottom_pad = 60

        normal_border = "#777777"
        empty_fill = "#fafafa"
        empty_border = "#cfcfcf"

        width = 420
        height = pad + 44 + room_gap + bottom_pad

        # Size the SVG to fit all rooms.
        for room in self.instance.rooms:
            room_width = room.getCols() * cell + room_inner_pad * 2
            room_height = title_h + room.getRows() * cell + room_inner_pad * 2
            width = max(width, pad * 2 + room_width)
            height += room_height + room_gap

        with open(file_path, "w", encoding="utf-8") as writer:
            writer.write(
                f'<svg xmlns="http://www.w3.org/2000/svg" width="{width}" height="{height}">\n'
            )

            writer.write("<style>\n")
            writer.write("text { font-family: Arial, sans-serif; fill: #000000; }\n")
            writer.write(".main-title { font-size: 24px; font-weight: 700; }\n")
            writer.write(".room-title { font-size: 20px; font-weight: 700; }\n")
            writer.write(".seat-student { font-size: 13px; font-weight: 700; }\n")
            writer.write(".seat-info { font-size: 12px; }\n")
            writer.write(".cheater-label { font-size: 11px; font-weight: 700; fill: #000000; }\n")
            writer.write(".room-card { fill: #ffffff; stroke: #dddddd; stroke-width: 1; }\n")
            writer.write("</style>\n")

            writer.write('<rect width="100%" height="100%" fill="#f7f8fa"/>\n')

            y = pad
            writer.write(
                f'<text x="{pad}" y="{y}" class="main-title">Exam seating visualization</text>\n'
            )
            y += 44 + room_gap

            for room in self.instance.rooms:
                room_card_x = pad
                room_card_y = y
                room_card_w = room.getCols() * cell + room_inner_pad * 2
                room_card_h = title_h + room.getRows() * cell + room_inner_pad * 2

                writer.write(
                    f'<rect x="{room_card_x}" y="{room_card_y}" '
                    f'width="{room_card_w}" height="{room_card_h}" '
                    f'rx="12" ry="12" class="room-card"/>\n'
                )

                writer.write(
                    f'<text x="{room_card_x + room_inner_pad}" '
                    f'y="{room_card_y + 30}" class="room-title">'
                    f'Room {room.getId()} ({room.getRows()}x{room.getCols()})</text>\n'
                )

                grid_x = room_card_x + room_inner_pad
                grid_y = room_card_y + title_h + room_inner_pad

                # layout[row][col] stores the student ID, or -1 if empty.
                layout = [[-1 for _ in range(room.getCols())] for _ in range(room.getRows())]

                # Convert seating to a room grid.
                for student_id, seat in enumerate(self.seating):
                    if (
                        seat.room_id == room.getId()
                        and 0 <= seat.row < room.getRows()
                        and 0 <= seat.col < room.getCols()
                    ):
                        layout[seat.row][seat.col] = student_id

                for row in range(room.getRows()):
                    for col in range(room.getCols()):
                        x = grid_x + col * cell
                        seat_y = grid_y + row * cell
                        student_id = layout[row][col]

                        if student_id == -1:
                            writer.write(
                                f'<rect x="{x}" y="{seat_y}" width="{cell}" height="{cell}" '
                                f'fill="{empty_fill}" stroke="{empty_border}" stroke-width="1"/>\n'
                            )
                        else:
                            student = self.instance.students[student_id]
                            group_fill = self._colorFromGroup(student.group_id)

                            writer.write(
                                f'<rect x="{x}" y="{seat_y}" width="{cell}" height="{cell}" '
                                f'fill="{group_fill}" stroke="{normal_border}" stroke-width="1"/>\n'
                            )

                            writer.write(
                                f'<text x="{x + cell // 2}" y="{seat_y + 25}" '
                                f'text-anchor="middle" class="seat-student">S{student.id}</text>\n'
                            )

                            writer.write(
                                f'<text x="{x + cell // 2}" y="{seat_y + 44}" '
                                f'text-anchor="middle" class="seat-info">G{student.group_id}</text>\n'
                            )

                            if student.cheater:
                                writer.write(
                                    f'<rect x="{x + 3}" y="{seat_y + 3}" '
                                    f'width="{cell - 6}" height="{cell - 6}" '
                                    f'fill="none" stroke="#000000" stroke-width="6"/>\n'
                                )

                                writer.write(
                                    f'<rect x="{x + 8}" y="{seat_y + 8}" '
                                    f'width="{cell - 16}" height="{cell - 16}" '
                                    f'fill="none" stroke="#ffffff" stroke-width="2"/>\n'
                                )

                                writer.write(
                                    f'<circle cx="{x + 14}" cy="{seat_y + 14}" r="10" '
                                    f'fill="#ffffff" stroke="#000000" stroke-width="2"/>\n'
                                )

                                writer.write(
                                    f'<text x="{x + 14}" y="{seat_y + 18}" '
                                    f'text-anchor="middle" class="cheater-label">C</text>\n'
                                )

                y += room_card_h + room_gap

            writer.write("</svg>\n")

File: Java/src/ExamSolution.java
import java.io.File;
import java.io.IOException;
import java.io.PrintStream;
import java.util.*;

/**
 * Represents one seating assignment. seating.get(i) is student i's seat.
 */
public class ExamSolution {
    ExamInstance instance;

    // seating.get(studentId) stores the room, row, and column assigned to that student.
    ArrayList<Seat> seating;

    /**
     * Creates an empty solution.
     */
    public ExamSolution(ExamInstance inst) {
        this.instance = inst;
        this.seating = new ArrayList<Seat>();

        ArrayList<Student> students = new ArrayList<>(instance.students);
        students.sort(Comparator.comparingInt(s -> s.groupId));

        ArrayList<ArrayList<Student>> groupedStudents = new ArrayList<>();
        for (Student student : students) {
            if (groupedStudents.get(student.groupId) != null) {
                groupedStudents.set(student.groupId, (ArrayList<Student>) List.of(new Student[]{student}));
            }
        }
    }

    /**
     * Creates a complete initial solution, optionally shuffled.
     */
    public ExamSolution(ExamInstance inst, boolean random) {
        this.instance = inst;
        this.seating = new ArrayList<Seat>();

        ArrayList<Seat> seats = new ArrayList<>();

        // Collect all available seats.
        for (Room room : instance.rooms) {
            for (int row = 0; row < room.getRows(); row++) {
                for (int col = 0; col < room.getCols(); col++) {
                    seats.add(new Seat(room.getId(), row, col));
                }
            }
        }

        if (random) Collections.shuffle(seats);

        // Assign seats to students 0..N-1.
        for (int i = 0; i < instance.N; i++) {
            seating.add(seats.get(i));
        }
    }

    /**
     * Returns an independent copy of this solution.
     */
    public ExamSolution copy() {
        ExamSolution sol = new ExamSolution(instance);

        for (Seat s : seating) {
            sol.seating.add(s.copy());
        }

        return sol;
    }

    /**
     * Assigns one student to one seat.
     */
    public void seatStudent(Student stud, Seat seat) {
        seating.set(stud.id, seat);
    }

    /**
     * Checks whether a seat is inside its room bounds.
     */
    private boolean isValidSeatPosition(Seat seat) {
        if (seat.getRoomId() < 0 || seat.getRoomId() >= instance.M) {
            return false;
        }

        Room room = instance.rooms.get(seat.getRoomId());

        return seat.getRow() >= 0 && seat.getRow() < room.getRows() &&
                seat.getCol() >= 0 && seat.getCol() < room.getCols();
    }

    /**
     * Checks that all students have valid, unique seats.
     */
    public boolean checkValid() {
        HashSet<Seat> taken = new HashSet<Seat>();

        for (int i = 0; i < instance.N; i++) {
            if (!isValidSeatPosition(seating.get(i))) return false;
            if (taken.contains(seating.get(i))) return false;
            taken.add(seating.get(i));
        }

        return true;
    }

    /**
     * Counts same-group students adjacent to listed cheaters.
     */
    public int cheaterIsolation() {
        int cheatPenalty = 0;

        for (Student s1 : instance.cheaters) {
            for (Student s2 : instance.groups.get(s1.groupId).students) {
                if (s1.id == s2.id) continue;

                Seat seat1 = seating.get(s1.id);
                Seat seat2 = seating.get(s2.id);

                if (seat1.roomId == seat2.roomId &&
                        Math.abs(seat1.row - seat2.row) <= 1 &&
                        Math.abs(seat1.col - seat2.col) <= 1) {
                    cheatPenalty++;
                }
            }
        }

        return cheatPenalty;
    }

    /**
     * Groups the seats of every exam group by room.
     */
    private TreeMap<Integer, ArrayList<Seat>> seatsByRoom(Group group) {
        TreeMap<Integer, ArrayList<Seat>> distr = new TreeMap<Integer, ArrayList<Seat>>();

        for (Student s : group.students) {
            Seat seat = seating.get(s.id);

            if (!distr.containsKey(seat.roomId)) {
                distr.put(seat.roomId, new ArrayList<Seat>());
            }

            distr.get(seat.roomId).add(seat);
        }

        return distr;
    }

    /**
     * Computes the group splitting part before normalization.
     */
    private double rawGroupSplittingCost() {
        double splittingCost = 0.0;

        for (Group group : instance.groups) {
            TreeMap<Integer, ArrayList<Seat>> distr = seatsByRoom(group);
            splittingCost += distr.size() * instance.cSplit;
        }

        return splittingCost;
    }

    /**
     * Computes the diameter part before normalization.
     */
    private double rawDiameterCost() {
        double diameterCost = 0.0;

        for (Group group : instance.groups) {
            TreeMap<Integer, ArrayList<Seat>> distr = seatsByRoom(group);

            for (ArrayList<Seat> A : distr.values()) {
                double diam = 0.0;

                // Find this group's diameter in the room.
                for (int i = 0; i < A.size(); i++) {
                    Seat seat1 = A.get(i);

                    for (int j = i + 1; j < A.size(); j++) {
                        Seat seat2 = A.get(j);

                        double distSquared =
                                (seat1.col - seat2.col) * (seat1.col - seat2.col) +
                                        (seat1.row - seat2.row) * (seat1.row - seat2.row);

                        diam = Math.max(diam, distSquared);
                    }
                }

                diameterCost += Math.sqrt(diam) * instance.cDiam;
            }
        }

        return diameterCost;
    }

    /**
     * Computes group splitting and diameter costs.
     */
    public double invigilatorCost() {
        return rawGroupSplittingCost() + rawDiameterCost();
    }

    /**
     * Computes the number of cheating penalties per room.
     */
    private int[] cheatPenaltyByRoom() {
        int[] penalties = new int[instance.M];

        for (Student s1 : instance.cheaters) {
            for (Student s2 : instance.groups.get(s1.groupId).students) {
                if (s1.id == s2.id) continue;

                Seat seat1 = seating.get(s1.id);
                Seat seat2 = seating.get(s2.id);

                if (seat1.roomId == seat2.roomId &&
                        Math.abs(seat1.row - seat2.row) <= 1 &&
                        Math.abs(seat1.col - seat2.col) <= 1) {
                    penalties[seat1.roomId]++;
                }
            }
        }

        return penalties;
    }

    /**
     * Formats doubles without unnecessary trailing zeroes.
     */
    private String formatDouble(double value) {
        if (Math.abs(value) < 1e-9) value = 0.0;

        return String.format(Locale.US, "%.6f", value)
                .replaceAll("0+$", "")
                .replaceAll("\\.$", "");
    }

    /**
     * Computes the normalized objective value.
     */
    public double getCost() {
        double cheatTerm = 0.0;

        if (!instance.cheaters.isEmpty()) {
            cheatTerm = (instance.cCheat / (double) instance.cheaters.size()) * cheaterIsolation();
        }

        return cheatTerm + (1.0 / (double) instance.groups.size()) * invigilatorCost();
    }

    /**
     * Writes a detailed cost breakdown to the statistics folder.
     * The file is written to statistics/<filename>.
     *
     */
    public void stats(String filename) {
        File file = new File("statistics/" + filename);

        int totalCheatPenalty = cheaterIsolation();
        int[] roomCheatPenalty = cheatPenaltyByRoom();

        double cheatingTerm = 0.0;
        if (!instance.cheaters.isEmpty()) {
            cheatingTerm = (instance.cCheat / (double) instance.cheaters.size()) * totalCheatPenalty;
        }

        double splittingCost = rawGroupSplittingCost();
        double diameterCost = rawDiameterCost();
        double invigilatorCost = splittingCost + diameterCost;
        double invigilatorTerm = 0.0;
        if (!instance.groups.isEmpty()) {
            invigilatorTerm = invigilatorCost / (double) instance.groups.size();
        }

        double totalCost = cheatingTerm + invigilatorTerm;

        try {
            PrintStream writer = new PrintStream(file);

            writer.println("Cheater Isolation Statistics");
            writer.println("----------------------------");
            writer.println("Total cheatPenalty = " + totalCheatPenalty);
            writer.println("Number of cheaters |Q| = " + instance.cheaters.size());
            writer.println("c_cheat = " + formatDouble(instance.cCheat));
            writer.println("Normalized cheating term = " + formatDouble(cheatingTerm));
            writer.println();

            writer.println("Cheat penalty by room:");
            for (int roomId = 0; roomId < instance.M; roomId++) {
                writer.println("Room " + roomId + " cheatPenalty = " + roomCheatPenalty[roomId]);
            }
            writer.println();

            writer.println("Invigilator Statistics");
            writer.println("----------------------");
            writer.println("Raw group splitting cost = " + formatDouble(splittingCost));
            writer.println("Raw diameter cost = " + formatDouble(diameterCost));
            writer.println("Raw invigilator cost = " + formatDouble(invigilatorCost));
            writer.println("Number of groups G = " + instance.groups.size());
            writer.println("Normalized invigilator term = " + formatDouble(invigilatorTerm));
            writer.println();

            writer.println("Combined Objective");
            writer.println("------------------");
            writer.println("Cheating term = " + formatDouble(cheatingTerm));
            writer.println("Invigilator term = " + formatDouble(invigilatorTerm));
            writer.println("Total cost F = " + formatDouble(totalCost));

            writer.close();

        } catch (IOException e) {
            System.out.println("Error: could not write statistics file");
            e.printStackTrace();
        }
    }

    /**
     * Writes one line per student: room row col.
     * The file is written to output/<filename>.
     *
     */
    public void output(String filename) {
        File file = new File(filename);

        try {
            PrintStream writer = new PrintStream(file);

            for (Seat seat : seating) {
                writer.println(seat.roomId + " " + seat.row + " " + seat.col);
            }

            writer.close();

        } catch (IOException e) {
            System.out.println("Error: file not found");
            e.printStackTrace();
        }
    }

    /**
     * Returns a color for a group.
     */
    private String colorFromGroup(int groupId) {
        String[] palette = {
                "#FFADAD",
                "#FFD6A5",
                "#FDFFB6",
                "#CAFFBF",
                "#9BF6FF",
                "#A0C4FF",
                "#BDB2FF",
                "#FFC6FF",
                "#B9FBC0",
                "#F1C0E8",
                "#CFBAF0",
                "#A3C4F3",
                "#90DBF4",
                "#98F5E1",
                "#FDE4CF",
                "#E4C1F9",
                "#D0F4DE",
                "#FCF6BD",
                "#FFCFD2",
                "#CDE7BE",
                "#FBC4AB",
                "#BDE0FE",
                "#CDB4DB",
                "#FFFFB5",
                "#D8F3DC"
        };

        if (groupId >= 0 && groupId < palette.length) {
            return palette[groupId];
        }

        // Fallback color for additional groups.
        float hue = (float) ((groupId * 0.618033988749895) % 1.0);
        float saturation = 0.42f;
        float brightness = 1.0f;

        int rgb = java.awt.Color.HSBtoRGB(hue, saturation, brightness);
        return String.format("#%06X", rgb & 0xFFFFFF);
    }

    /**
     * Writes an SVG visualization of the seating plan.
     * <p>
     * The file is written to visualizations/<filename>.
     */
    public void visualize(String filename) {
        final int cell = 64;
        final int pad = 32;
        final int titleH = 44;
        final int roomGap = 42;
        final int roomInnerPad = 18;
        final int bottomPad = 60;

        final String normalBorder = "#777777";
        final String emptyFill = "#fafafa";
        final String emptyBorder = "#cfcfcf";

        final int normalBorderWidth = 1;
        final int emptyBorderWidth = 1;

        int width = 420;
        int height = pad + 44 + roomGap + bottomPad;

        // Size the SVG to fit all rooms.
        for (Room room : instance.rooms) {
            int roomWidth = room.getCols() * cell + roomInnerPad * 2;
            int roomHeight = titleH + room.getRows() * cell + roomInnerPad * 2;

            width = Math.max(width, pad * 2 + roomWidth);
            height += roomHeight + roomGap;
        }

        File file = new File("visualizations/" + filename);

        try {
            PrintStream writer = new PrintStream(file);

            writer.println("<svg xmlns=\"http://www.w3.org/2000/svg\" width=\"" + width + "\" height=\"" + height + "\">");

            writer.println("<style>");
            writer.println("text { font-family: Arial, sans-serif; fill: #000000; }");
            writer.println(".main-title { font-size: 24px; font-weight: 700; }");
            writer.println(".room-title { font-size: 20px; font-weight: 700; }");
            writer.println(".seat-student { font-size: 13px; font-weight: 700; }");
            writer.println(".seat-info { font-size: 12px; }");
            writer.println(".cheater-label { font-size: 11px; font-weight: 700; fill: #000000; }");
            writer.println(".room-card { fill: #ffffff; stroke: #dddddd; stroke-width: 1; }");
            writer.println("</style>");

            writer.println("<rect width=\"100%\" height=\"100%\" fill=\"#f7f8fa\"/>");

            int y = pad;

            writer.println("<text x=\"" + pad + "\" y=\"" + y + "\" class=\"main-title\">Exam seating visualization</text>");
            y += 44 + roomGap;

            for (Room room : instance.rooms) {
                int roomCardX = pad;
                int roomCardY = y;
                int roomCardW = room.getCols() * cell + roomInnerPad * 2;
                int roomCardH = titleH + room.getRows() * cell + roomInnerPad * 2;

                writer.println(
                        "<rect x=\"" + roomCardX +
                                "\" y=\"" + roomCardY +
                                "\" width=\"" + roomCardW +
                                "\" height=\"" + roomCardH +
                                "\" rx=\"12\" ry=\"12\" class=\"room-card\"/>"
                );

                writer.println(
                        "<text x=\"" + (roomCardX + roomInnerPad) +
                                "\" y=\"" + (roomCardY + 30) +
                                "\" class=\"room-title\">Room " + room.getId() +
                                " (" + room.getRows() + "x" + room.getCols() + ")</text>"
                );

                int gridX = roomCardX + roomInnerPad;
                int gridY = roomCardY + titleH + roomInnerPad;

                // layout[row][col] stores the student ID, or -1 if empty.
                int[][] layout = new int[room.getRows()][room.getCols()];

                for (int row = 0; row < room.getRows(); row++) {
                    Arrays.fill(layout[row], -1);
                }

                // Convert seating to a room grid.
                for (int studentId = 0; studentId < seating.size(); studentId++) {
                    Seat seat = seating.get(studentId);

                    if (seat.roomId == room.getId()
                            && seat.row >= 0 && seat.row < room.getRows()
                            && seat.col >= 0 && seat.col < room.getCols()) {
                        layout[seat.row][seat.col] = studentId;
                    }
                }

                for (int row = 0; row < room.getRows(); row++) {
                    for (int col = 0; col < room.getCols(); col++) {
                        int x = gridX + col * cell;
                        int seatY = gridY + row * cell;
                        int studentId = layout[row][col];

                        if (studentId == -1) {
                            writer.println(
                                    "<rect x=\"" + x +
                                            "\" y=\"" + seatY +
                                            "\" width=\"" + cell +
                                            "\" height=\"" + cell +
                                            "\" fill=\"" + emptyFill +
                                            "\" stroke=\"" + emptyBorder +
                                            "\" stroke-width=\"" + emptyBorderWidth +
                                            "\"/>"
                            );
                        } else {
                            Student student = instance.students.get(studentId);
                            String groupFill = colorFromGroup(student.groupId);

                            writer.println(
                                    "<rect x=\"" + x +
                                            "\" y=\"" + seatY +
                                            "\" width=\"" + cell +
                                            "\" height=\"" + cell +
                                            "\" fill=\"" + groupFill +
                                            "\" stroke=\"" + normalBorder +
                                            "\" stroke-width=\"" + normalBorderWidth +
                                            "\"/>"
                            );

                            writer.println(
                                    "<text x=\"" + (x + cell / 2) +
                                            "\" y=\"" + (seatY + 25) +
                                            "\" text-anchor=\"middle\" class=\"seat-student\">S" +
                                            student.id + "</text>"
                            );

                            writer.println(
                                    "<text x=\"" + (x + cell / 2) +
                                            "\" y=\"" + (seatY + 44) +
                                            "\" text-anchor=\"middle\" class=\"seat-info\">G" +
                                            student.groupId + "</text>"
                            );

                            boolean isCheater = instance.cheaters.contains(student);

                            if (isCheater) {
                                writer.println(
                                        "<rect x=\"" + (x + 3) +
                                                "\" y=\"" + (seatY + 3) +
                                                "\" width=\"" + (cell - 6) +
                                                "\" height=\"" + (cell - 6) +
                                                "\" fill=\"none\" stroke=\"#000000\" stroke-width=\"6\"/>"
                                );

                                writer.println(
                                        "<rect x=\"" + (x + 8) +
                                                "\" y=\"" + (seatY + 8) +
                                                "\" width=\"" + (cell - 16) +
                                                "\" height=\"" + (cell - 16) +
                                                "\" fill=\"none\" stroke=\"#ffffff\" stroke-width=\"2\"/>"
                                );

                                writer.println(
                                        "<circle cx=\"" + (x + 14) +
                                                "\" cy=\"" + (seatY + 14) +
                                                "\" r=\"10\" fill=\"#ffffff\" stroke=\"#000000\" stroke-width=\"2\"/>"
                                );

                                writer.println(
                                        "<text x=\"" + (x + 14) +
                                                "\" y=\"" + (seatY + 18) +
                                                "\" text-anchor=\"middle\" class=\"cheater-label\">C</text>"
                                );
                            }
                        }
                    }
                }

                y += roomCardH + roomGap;
            }

            writer.println("</svg>");
            writer.close();

        } catch (IOException e) {
            System.out.println("Error: could not write visualization file");
            e.printStackTrace();
        }
    }
}